crates/datasources/src/clickhouse/stream.rs

```diff
@@ -0,0 +1,244 @@
+}
+
+/// Converts a column from a block into an arrow array.
+///
+/// The column's data type should be known beforehand.
+fn column_to_array(
+    datatype: DataType,
+    column: &Column<Simple>,
+    nullable: bool,
+) -> Result<Arc<dyn Array>> {
+    // TODO: This could be a function, but I'm not too keen on figuring out the
+    // types right now.
+    macro_rules! make_primitive_array {
+        ($primitive:ty, $arr_type:ty, $nullable:expr) => {{
+            if nullable {
+                let vals: Vec<_> = column
+                    .iter::<Option<$primitive>>()?
+                    .map(|opt| opt.cloned())
+                    .collect();
+                Arc::new(<$arr_type>::from(vals))
+            } else {
+                let vals: Vec<_> = column.iter::<$primitive>()?.cloned().collect();
+                Arc::new(<$arr_type>::from(vals))
+            }
+        }};
+    }
+
+    let arr: Arc<dyn Array> = match datatype {
+        DataType::Boolean => make_primitive_array!(bool, BooleanArray, nullable),
+        DataType::UInt8 => make_primitive_array!(u8, UInt8Array, nullable),
+        DataType::UInt16 => make_primitive_array!(u16, UInt16Array, nullable),
+        DataType::UInt32 => make_primitive_array!(u32, UInt32Array, nullable),
+        DataType::UInt64 => make_primitive_array!(u64, UInt64Array, nullable),
+        DataType::Int8 => make_primitive_array!(i8, Int8Array, nullable),
+        DataType::Int16 => make_primitive_array!(i16, Int16Array, nullable),
+        DataType::Int32 => make_primitive_array!(i32, Int32Array, nullable),
+        DataType::Int64 => make_primitive_array!(i64, Int64Array, nullable),
+        DataType::Float32 => make_primitive_array!(f32, Float32Array, nullable),
+        DataType::Float64 => make_primitive_array!(f64, Float64Array, nullable),
+        DataType::Utf8 => {
+            if nullable {
+                let vals: Vec<_> =
+                    <Option<&[u8]> as Iterable<Simple>>::iter(column, column.sql_type())?
+                        .map(|bs| bs.map(|bs| str::from_utf8(bs).unwrap()))
+                        .collect();
+                Arc::new(StringArray::from(vals))
+            } else {
+                let vals: Vec<_> = <&[u8]>::iter(column, column.sql_type())?
+                    .map(|bs| str::from_utf8(bs).unwrap())
+                    .collect();
+                Arc::new(StringArray::from(vals))
+            }
+        }
+        DataType::Date32 => {
+            let epoch = NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
+
+            if nullable {
+                let vals: Vec<_> =
+                    <Option<NaiveDate> as Iterable<Simple>>::iter(column, column.sql_type())?
+                        .collect();
+                Arc::new(Date32Array::from(
+                    vals.into_iter()
+                        .map(|date| {
+                            date.map(|date| date.signed_duration_since(epoch).num_days() as i32)
+                        })
```

Comment:
why not do this in a single `.iter()` instead of collecting only to run `into_iter` immediately afterwards?

Reply:
Oversight, I was busy playing type jenga. Fixed in latest commit.